suite/tests/CMakeLists.txt

```diff
@@ -553,6 +553,24 @@ function(add_exe test source)
       "${CMAKE_GENERATOR}" MATCHES "Visual Studio")
     add_dependencies(${test} ${gen_asm_tgt})
   endif ()
+  if (BUILD_TESTS_SVE)
+    target_compile_options(${test} PRIVATE
+      "-march=armv8.4-a+crypto+rcpc+sha3+sm4+sve+rng+ssbs+nodotprod")
+
+    # Some tests give very obscure build errors when built with "-O3"
+    # FIXME i#5365: fix build issues due to optimisation, unless no
+    # optimisation is a requirement for a test case.
```

Comment:
I am a little confused: aren't these tests already built -O3 in a release build (see above about expecting tests to match the rest of the project), which is a build setting in the GA CI suite?  So we should already be exercising -O3, so is it the SVE flags causing the problems you are seeing?

Despite thinking we already exercise -O3, it is not surprising to see some tests fail with high optimization as they often construct particular patterns using simple conditions and often have unused results or potentially-dead code.  This is esp true for tool tests like Dr. Memory (in a different repository) which wants weird and buggy code patterns and often has to fight the compiler.  You could image drcov or other tools having similar problems.  You could imagine drmemtrace tests having load/store references for testing without significant use and the compiler throws them away at high optimization.  But again I thought we already built these -O3.

Reply:
OK I found on line 280 of this file:

```
  # some tests rely on specific "nop;nop" patterns that optimization ruins
  # we should probably move the -O from top level into core/CMakeLists.txt
  string(REGEX REPLACE "-O[0-9]? " " " CMAKE_C_FLAGS "${CMAKE_C_FLAGS}")
```

Some client tests have the nops as sentinels.  If annotations were implemented on aarch64 (#1672) we could probably switch to those in all cases but we don't have that today.

So the tests, in in release build, are in fact built without high optimizations today.

Given the nop issues: probably for the short term only the tests you are targeting with auto-vectorization should have -O3 enabled; i.e., an allowlist, instead of the denylist here.  Have you run all the tests in -O3?  According to that comment we'd expect many client tests to fail.  Was the not the case?

Reply:
Or, we can try to change that line 280 removal of -O* to only remove for a list of tests that have the nop;nop or similar patterns.

Reply:
With -O3 we get 36 test failures rather than the normal 6. But most of the tests still pass.

I think the plan is to investigate the new failures and if necessary build them without -O3

Reply:
I like your idea of removing line 280 and removing -O3 in a more granular way. We can look into that.

Reply:
I looked into removing the line
`string(REGEX REPLACE "-O[0-9]? " " " CMAKE_C_FLAGS "${CMAKE_C_FLAGS}")`
and selectively adding the `-O0` flag to certain sources using `target_compile_options`

However this didn't work as the files were built with both `-O0` and `-O3` flags but the `-O3` came last so had precedence.

We could have an allow list for adding `-O3` but it would be much longer than the deny list.

Reply:
Removing compile flags in CMake in a local manner has never been well-supported, which is why we have such hacky solutions like clearing CMAKE_C*_FLAGS completely in core DR so we can control it fully (which then leads to other problems like missing debug flags on "regular" targets).  I haven't kept up with new CMake features (I had never heard of `target_compile_options` until this PR): is it possible there is new support in CMake for removing options locally (per-directory or per-target) or for specifying that our target flags should be appended instead of prepended to the base flags?

Reply:
@derekbruening 
My approach at the moment is to remove the line
`string(REGEX REPLACE "-O[0-9]? " " " CMAKE_C_FLAGS "${CMAKE_C_FLAGS}")`
and then add code in `set_cflags` (for core tests) and `_DR_set_compile_flags` (for client tests) to remove "-O3" for tests that fail when built this way.

Although my change will end up being more verbose than the original change, it will avoid the blanket removal of "-Ox" for all tests.

Does this approach sound reasonable?

Reply:
`_DR_set_compile_flags` in `make/DynamoRIOConfig.cmake.in` is used for real user clients, not just for tests, so that is not the right place to remove flags as users will not be happy with that.  That would presumably include us as that would remove optimizations from all of our provided tools including drmemtrace, drcachesim, etc. in our release packages.

Reply:
@derekbruening I have reworked this PR now to detect if the build machine supports SVE rather than use a build option. I also have a whitelist of tests to build for SVE rather than a blacklist of tests not to build.

Reply:
Please update the description which still talks about a build option.  If you click on the circular arrow you can then re-request a review.